File: controllers/url.js
const shortid = require("shortid");
const { getDB } = require("../db.js");

// Create new short URL
async function handleGeneratedNewShortUrl(req, res) {
    const body = req?.body;
    if (!body?.url) return res.status(400).json({ error: 'url is required' });

    const shortID = shortid.generate();
    const redirectURL = body.url;
    let client;

    try {
        const pool = await getDB();
        client = await pool.connect();
        console.log("Acquired PG connection for inserting URL");

        const insertQuery = `
            INSERT INTO urls (shortId, redirectURL)
            VALUES ($1, $2)
            RETURNING id
        `;
        const result = await client.query(insertQuery, [shortID, redirectURL]);

        console.log("Inserted URL with ID:", result.rows[0].id);
        return res.status(201).json({ message: "Short URL created!", id: shortID });

    } catch (error) {
        console.error("Error saving URL:", error);
        // Handle duplicate shortId
        if (error.code === '23505') { // PostgreSQL unique violation
            return res.status(409).json({ error: 'Short ID already exists. Please try again.' });
        }
        return res.status(500).json({ error: 'Failed to create short URL' });
    } finally {
        if (client) {
            client.release();
            console.log("Released PG connection after inserting URL");
        }
    }
}

// Get analytics for a short URL
async function handleGetAnalytics(req, res) {
    const shortId = req.params.shortId;
    if (!shortId) return res.status(400).json({ error: 'Short ID is required' });

    let client;

    try {
        const pool = await getDB();
        client = await pool.connect();
        console.log("Acquired PG connection for analytics");

        // Find the URL first
        const findUrlQuery = `
            SELECT id, shortId, redirectURL, createdAt, updatedAt
            FROM urls
            WHERE shortId = $1
        `;
        const urlResult = await client.query(findUrlQuery, [shortId]);

        if (urlResult.rows.length === 0) {
            return res.status(404).json({ error: 'Short URL not found' });
        }

        const urlData = urlResult.rows[0];

        // Get visit history
        const findHistoryQuery = `
            SELECT visit_timestamp
            FROM visit_history
            WHERE url_id = $1
            ORDER BY visit_timestamp DESC
        `;
        const historyResult = await client.query(findHistoryQuery, [urlData.id]);

        const analytics = historyResult.rows.map(row => ({ timestamp: row.visit_timestamp }));

        return res.json({
            ShortID: urlData.shortid,
            realUrlLink: urlData.redirecturl,
            CreateUrlTime: urlData.createdat,
            UpdateUrlTime: urlData.updatedat,
            totalClicks: historyResult.rows.length,
            analytics: analytics,
        });

    } catch (error) {
        console.error("Error getting analytics:", error);
        return res.status(500).json({ error: 'Failed to retrieve analytics' });
    } finally {
        if (client) {
            client.release();
            console.log("Released PG connection after getting analytics");
        }
    }
}

module.exports = {
    handleGeneratedNewShortUrl,
    handleGetAnalytics,
};

File: controllers/url.py
from flask import request, jsonify
from db import get_db
import shortuuid

def handle_generated_new_short_url():
    body = request.get_json()
    if not body or not body.get('url'):
        return jsonify({'error': 'url is required'}), 400

    short_id = shortuuid.uuid()[:8]  # Generate short ID
    redirect_url = body['url']
    client = None

    try:
        pool = get_db()
        client = pool.getconn()
        print("Acquired connection for inserting URL")

        with client.cursor() as cursor:
            insert_query = 'INSERT INTO urls (shortid, redirecturl) VALUES (%s, %s) RETURNING id'
            cursor.execute(insert_query, (short_id, redirect_url))
            result = cursor.fetchone()
            client.commit()

            print(f"Inserted URL with ID: {result[0]}")
            return jsonify({'message': 'Short URL created!', 'id': short_id}), 201

    except Exception as error:
        print(f"Error saving URL: {error}")
        if client:
            client.rollback()
        
        # Check for duplicate entry error
        if 'duplicate key' in str(error).lower() or 'unique constraint' in str(error).lower():
            return jsonify({'error': 'Short ID already exists. Please try again.'}), 409
        return jsonify({'error': 'Failed to create short URL'}), 500
    finally:
        if client:
            print("Releasing connection after inserting URL")
            pool.putconn(client)

def handle_get_analytics(short_id):
    if not short_id:
        return jsonify({'error': 'Short ID is required'}), 400

    client = None
    try:
        pool = get_db()
        client = pool.getconn()
        print("Acquired connection for getting analytics")

        with client.cursor() as cursor:
            # Find the URL first
            find_url_query = 'SELECT id, shortid, redirecturl, createdat, updatedat FROM urls WHERE shortid = %s'
            cursor.execute(find_url_query, (short_id,))
            url_row = cursor.fetchone()

            if not url_row:
                return jsonify({'error': 'Short URL not found'}), 404

            url_id, short_id_db, redirect_url, created_at, updated_at = url_row

            # Get visit history
            find_history_query = 'SELECT visit_timestamp FROM visit_history WHERE url_id = %s ORDER BY visit_timestamp DESC'
            cursor.execute(find_history_query, (url_id,))
            history_rows = cursor.fetchall()

            analytics = [{'timestamp': str(row[0])} for row in history_rows]

            return jsonify({
                'ShortID': short_id_db,
                'realUrlLink': redirect_url,
                'CreateUrlTime': str(created_at) if created_at else None,
                'UpdateUrlTime': str(updated_at) if updated_at else None,
                'totalClicks': len(history_rows),
                'analytics': analytics,
            })

    except Exception as error:
        print(f"Error getting analytics: {error}")
        return jsonify({'error': 'Failed to retrieve analytics'}), 500
    finally:
        if client:
            print("Releasing connection after getting analytics")
            pool.putconn(client)

